packages/template-retail-react-app/app/components/product-view/index.jsx

```diff
@@ -141,6 +142,64 @@ const ProductView = forwardRef(
         } = useBonusProductModalContext()
         const theme = useTheme()
         const [showOptionsMessage, toggleShowOptionsMessage] = useState(false)
+        const [promotionIdToSearch, setPromotionIdToSearch] = useState(null)
+        const {data: bonusProductSearchResult} = useBonusProductSearch(promotionIdToSearch)
+
+        // State to track all promotion IDs and their results
+        const [pendingPromotionIds, setPendingPromotionIds] = useState([])
+        const [bonusItemsForModal, setBonusItemsForModal] = useState(null)
+        const ruleBasedPromotionsRef = useRef([])
+
+        // Effect to handle multiple promotions sequentially
+        useEffect(() => {
+            if (bonusProductSearchResult?.hits?.length > 0 && promotionIdToSearch) {
+                // Format the bonus products for current promotion
+                let formattedBonusProducts = bonusProductSearchResult.hits.map((bonusProduct) => {
+                    return {
+                        productId: bonusProduct.productId,
+                        productName: bonusProduct.productName,
+                        c_productUrl: bonusProduct.c_productUrl
+                    }
```

Comment:
If this is the data structure we are dealing for bonus product. You can avoid doing this here by taking advantage of react-query `select` data transformation during useBonusProductSearch hook. then the data returns will always have this shape.

https://tanstack.com/query/v4/docs/framework/react/reference/useQuery
See `useProductViewModal` for an example